GS/HW: Remove handling of provoking first vertex in early pipeline.

TJnotJT opened the pull request:
### Description of Changes
Remove the handling of provoking first vertex in early pipeline GSState and GSVertexTrace (and other helper classes) and move it later pipeline GSRendererHW. This is mainly for DX11-12, Vulkan (when it doesn't support fast vertex provoking), and Metal. It should not affect OGL.

### Rationale behind Changes
There is unnecessary code complexity added early in the pipeline in GSState/GSVertexTrace (e.g., keeping track of when vertices are swapped in primitives and when not). Since provoking first vertex is a graphics API quirk, it is better to handle it in lower level code in GSRendererHW.

Also, the way provoking first vertex was implemented gave incorrect results in some cases since it swapped vertices in the draw. This could cause incorrect drawing of line primitives since lines may (and are usually) not drawn the same when interchanging the two endpoints. Affects DX11-12 and possibly Vulkan/Metal. See examples below.

Finally, the code paths that would actually be affected by the provoking first vertex behavior seem to rare (i.e., it only applies to flat shaded triangles/lines, among other criteria) so there is no point in handling it early in the pipeline (even though it may be cheaper to do so). Handling it later may have a small perf impact because the entire vertex buffer may need to be copied. However, I did a simple analysis of the ~2000 GS dumps in the testing collection and it seems that on average < 0.5% of draws actually use this path.

### Suggested Testing Steps
Tested this by doing a GS dump run with DX11. I have not tested on upscaling so that would be very helpful (must be tested eventually because upscaling shaders in DX11-12 were modified). Comparing between DX11-12/Vulkan/Metal and OGL could be helpful since OGL should not be affected by this PR. Testing random games would also work.

Some fixes for DX11 (PR left, master right):

![007](https://github.com/user-attachments/assets/a92be573-06f7-4f85-83ce-bcf5bedcaa85)
GS dump: "007 - Everything or Nothing_SLUS-20751_20240622211641.gs.xz"

![thx](https://github.com/user-attachments/assets/c0e00651-271b-4301-adc1-338a857e5a98)
GS dump from this issue: https://github.com/PCSX2/pcsx2/issues/12825.

### Did you use AI to help find, test, or implement this issue or feature?
GitHub Copilot.

TJnotJT commented on pcsx2/GS/GSState.cpp:
```diff
@@ -3856,18 +3851,18 @@ __forceinline void GSState::VertexKick(u32 skip)
 			m_index.tail += 3;
 			break;
-		case GS_TRIANGLEFAN:
+		case GS_TRIANGLEFAN:		
```
Thanks, fixed

TellowKrinkle commented on pcsx2/GS/Renderers/HW/GSRendererHW.cpp:
```diff
@@ -4782,6 +4775,81 @@ bool GSRendererHW::VerifyIndices()
+	}
+	else
+	{
+		// Reallocate the vertex buffer
+		m_vertex.maxcount = std::max(m_vertex.maxcount, m_index.tail);
+		GSVertex* vert_buff = static_cast<GSVertex*>(_aligned_malloc(sizeof(GSVertex) * m_vertex.maxcount, 32));
+
+		// De-index and copy the vertices
+		for (u32 i = 0; i < m_index.tail; i++)
+		{
+			vert_buff[i] = m_vertex.buff[m_index.buff[i]];
+			m_index.buff[i] = static_cast<u16>(i);
+		}
+		std::swap(vert_buff, m_vertex.buff);
+		_aligned_free(vert_buff);
```
We should probably have a secondary vertex buffer that gets kept at the same size as the main one, so any time you want to rewrite vertices you can just do it from one to the other and then swap them, and then you won't have to worry about fans being sneaky.

refractionpcsx2 replied:
We kinda do have a secondary one, but right now we only use it for buffered channel shuffles

TJnotJT replied:
@TellowKrinkle 
Agreed, having a secondary buffer will reduce some of the reallocation/deallocation and be more correct. Made the changes and pushed.

@refractionpcsx2 
Yeah, I noticed that buffer. Hopefully there won't be any bad interactions with the code I added since that buffer only copied in SetupIA() after the vertex colors are fixed by HandleProvokingFirstVertex().

lightningterror commented:
Is the temp commit still needed, so we can get this tested and merged?

TJnotJT commented:
> Is the temp commit still needed, so we can get this tested and merged?

Thanks for checking. Removed the debugging code and rebased so should be all set to test. So far I've done dump runs on DX11, Vulkan. I will check SW in case in next couple days.

TJnotJT commented:
Forgot to remove one line of debugging code that prevented compilation. Should be fixed now.

lightningterror reviewed (commented):
Still need to bump shader cache
https://github.com/PCSX2/pcsx2/blob/8937fb3d11c0d9f1faa3ccd5e185375b1cc71ad1/pcsx2/ShaderCacheVersion.h#L6

TJnotJT commented:
Apologies, I didn't understand that before. Though it meant to update the Vulkan shader itself. Just updated the shader cache.

TJnotJT commented:
SW dump run looks good. There's an issue with at least one dump in OGL that I need to look into.

TJnotJT commented:
Those involved already know this, but for anyone else the above Baldur's Gate issue may be a regression noted here https://github.com/PCSX2/pcsx2/issues/13025 and here https://github.com/PCSX2/pcsx2/pull/13026

lightningterror commented:
You can rebase this since I merged the regression fix.

TJnotJT commented:
Yep, all done.

lightningterror commented on pcsx2/GS/GSState.h:
```diff
@@ -137,6 +136,7 @@ class GSState : public GSAlignedClass<32>
 	struct
 	{
 		GSVertex* buff;
+		GSVertex* buff2; // same size buffer to copy/modify the original buffer without constantly reallocating
```
Maybe naming it temp_buff, or something similar would be more appropriate.

TJnotJT replied:
How about swap_buff? In reality it's not really a temp or scratch buffer so my original commit message wasn't accurate.

lightningterror commented on pcsx2/GS/GSState.cpp:
```diff
@@ -2865,56 +2861,52 @@ void GSState::UpdateVertexKick()
+	std::vector<AllocDesc> alloc_desc = {
+		{reinterpret_cast<void**>(&m_vertex.buff),      old_vertex_size, new_vertex_size},
+		{reinterpret_cast<void**>(&m_vertex.buff_copy),  0,               new_vertex_size}, // discard contents of buff_copy
+		{reinterpret_cast<void**>(&m_draw_vertex.buff),  old_vertex_size, new_vertex_size},
```
Extra space before old_vertex_size, same for the above.

lightningterror commented on pcsx2/GS/Renderers/HW/GSRendererHW.cpp:
```diff
@@ -4818,6 +4807,52 @@ bool GSRendererHW::VerifyIndices()
+void GSRendererHW::HandleProvokingVertexFirst()
+{
+	// Early exit conditions:
+	if (g_gs_device->Features().provoking_vertex_last ||	// device supports provoking last vertex
```
The comments here can be better formatted too (including below ones).

lightningterror commented:
Did a smoke test on all renderers on the affected dumps and seems fine, also tellow says pr looks good.
